crypto/digest_extra/digest_test.cc

```diff
@@ -527,44 +528,176 @@ TEST(DigestTest, TransformBlocks) {
     uint8_t hash2[HASH_NAME##_DIGEST_LENGTH];                                 \
     EXPECT_TRUE(HASH_NAME##_Final(hash2, &ctx3)) << "final 3";                \
                                                                               \
+    /* finally check the resulting hash matches the baseline hash */          \
     EXPECT_EQ(Bytes(hash1), Bytes(hash2)) << "same hash";                     \
   }
 
-// The macro DIGEST_TEST_InitAndGetState_Body only contains the body of the
-// tests and not the TEST(...) { part, so that IDE can still easily recognize
-// those as tests.
 
-TEST(DigestTest, InitAndGetStateMD5) {
-  DIGEST_TEST_InitAndGetState_Body(MD5, MD5_CTX, MD5_CBLOCK);
+// DIGEST_TEST_InitAndGetStateLarge_Body expands to the body of the various
+// InitAndGetStateLarge* tests below.
+// Compared to DIGEST_TEST_InitAndGetStateBasic_Body, it allows testing
+// hashing an arbitrary amount of data.
+// Of particular interest is the choice of NB_1000_BLOCKS to be such that
+// the number of hash bits n > 2^32, therefore overflowing uint32_t (used for
+// SHA-1/SHA-256 but not SHA-512 which uses uint64_t).
+// Concretely, the test evaluates the hash function over (NB_1000_BLOCKS+2)*1000
+// blocks, exporting the state after NB_1000_BLOCKS * 1000 blocks.
+// Because EXPECT_* outputs the line where the macro is called/expanded, we need
+// to add diagnostic information `<< ...` after each EXPECT_*
+#define DIGEST_TEST_InitAndGetStateLarge_Body(HASH_NAME, HASH_CTX,             \
+                                              HASH_CBLOCK, NB_1000_BLOCKS)     \
+  {                                                                            \
+    const size_t block_size = HASH_CBLOCK;                                     \
+    uint8_t data_1000_blocks[block_size * 1000];                               \
+    for (size_t i = 0; i < sizeof(data_1000_blocks); i++) {                    \
+      data_1000_blocks[i] = i * 3;                                             \
+    }                                                                          \
+                                                                               \
+    /* Compute the hash of the NB_1000_BLOCKS+2 times the data_1000_blocks for \
+     * the baseline */                                                         \
+    HASH_CTX ctx1;                                                             \
+    EXPECT_TRUE(HASH_NAME##_Init(&ctx1)) << "init 1";                          \
+    for (size_t i = 0; i < NB_1000_BLOCKS + 2; i++) {                          \
+      EXPECT_TRUE(HASH_NAME##_Update(&ctx1, data_1000_blocks,                  \
+                                     sizeof(data_1000_blocks)))                \
+          << "update 1 - i=" << i;                                             \
```

Comment:
I don't remember if between quotes of a string you can concatenate `HASH_NAME##` so that the output messages are more specific (here and the macro above).

Reply:
Because the macro is expanded into the actual test function, we already can know where it comes from (as the function name is included in the output).
Concretely, here is an actual output when forcing an error:

```
$ ./crypto/crypto_test --gtest_filter='DigestTest.InitAndGetState*'
Note: Google Test filter = DigestTest.InitAndGetState*
[==========] Running 16 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 16 tests from DigestTest
[ RUN      ] DigestTest.InitAndGetStateMD5Basic
[...]/digest_test.cc:624: Failure
Value of: MD5_get_state(&ctx2, state_h, &state_n)
  Actual: true
Expected: false
get_state 2b
[  FAILED  ] DigestTest.InitAndGetStateMD5Basic (0 ms)
```